### ISSUE TITLE
Calculating gradient only on smooth penalty in train_model_ista for cLSTM

### ISSUE THREAD

**Issue by tsnnay** [icc2115/Neural-GC#7]
I am wondering the reason you are calculating gradient only on smooth loss in the `train_model_ista` function, even when we are making the proximal gradient step.

My doubt is shouldn't we call `joint_loss.backward()`, where `joint_loss=smooth + nonsmooth`? And this might be a little trivial, but can you explain the reasoning for taking `mean_loss` by dividing the 'smooth + nonsmooth' term with no. of features

**Comment by iancovert** [comment]
Hi, yes happy to explain. When you're using ISTA, aka proximal gradient descent, it's because you're optimizing a non-smooth objective function. You can split the objective function into two terms, as you've indicated with `joint_loss = smooth + nonsmooth`. And the way ISTA works is that you alternate taking a gradient step on the smooth part and then taking a proximal step for the non-smooth part.

If you want to just call `joint_loss.backward()`, you can do that and optimize with SGD or Adam - but then you should skip the proximal step. We have results for that in the most recent version of our paper - it works just as well, the only problem is you won't get exact sparsity. Using ISTA is a way to guarantee exact sparsity. 

Dividing `mean_loss` by the number of features just makes the loss more presentable to the user, in case the time series has many dimensions. This could be omitted and it wouldn't affect training.
